source/base/parameter_handler.cc

```diff
@@ -1837,78 +1812,50 @@ namespace
-                                      path_separator,
-                                      patterns,
-                                      destination);
-
-            // see if the recursive read
-            // succeeded. if yes, continue,
-            // otherwise exit now
-            if (result == false)
-              return false;
+            read_xml_recursively (p->second,
+                                  (current_path == "" ?
+                                   p->first :
+                                   current_path + path_separator + p->first),
+                                  path_separator,
+                                  patterns,
+                                  destination);
           }
       }
-
-    return true;
   }
 }
 
 
 
-bool ParameterHandler::read_input_from_xml (std::istream &in)
+void ParameterHandler::read_input_from_xml (std::istream &in)
 {
   AssertThrow(in, ExcIO());
   // read the XML tree assuming that (as we
   // do in print_parameters(XML) it has only
   // a single top-level node called
   // "ParameterHandler"
   boost::property_tree::ptree single_node_tree;
-  try
-    {
-      read_xml (in, single_node_tree);
-    }
-  catch (...)
-    {
-      std::cerr << "This input stream appears not to be valid XML"
-                << std::endl;
-      return false;
-    }
+  // This boost function will raise an exception if this is not a valid XML
+  // file.
+  read_xml (in, single_node_tree);
 
-  // make sure there is a top-level element
+  // make sure there is a single top-level element
   // called "ParameterHandler"
-  if (!single_node_tree.get_optional<std::string>("ParameterHandler"))
-    {
-      std::cerr << "There is no top-level XML element called \"ParameterHandler\"."
-                << std::endl;
-      return false;
-    }
+  AssertThrow (single_node_tree.get_optional<std::string>("ParameterHandler"),
+               ExcMessage ("There is no top-level XML element called "
+                           "\"ParameterHandler\"."));
 
-  // ensure that there is only a single
-  // top-level element
-  if (std::distance (single_node_tree.begin(), single_node_tree.end()) != 1)
-    {
-      std::cerr << "The top-level XML element \"ParameterHandler\" is "
-                << "not the only one."
-                << std::endl;
-      std::cerr << "(There are "
-                << std::distance (single_node_tree.begin(),
-                                  single_node_tree.end())
-                << " top-level elements.)"
-                << std::endl;
-      return false;
-    }
+  const std::size_t n_top_level_elements = std::distance (single_node_tree.begin(),
+                                                          single_node_tree.end());
+  AssertThrow (n_top_level_elements == 1,
+               ExcMessage ("There are "
+                           + Utilities::to_string(n_top_level_elements)
+                           + " top-level elements, but there should only be one."));
```

Comment:
I think it would be useful to also state *of what* there are more than one top-level elements.